### PR TITLE
add: logging for debugging

### PR DESCRIPTION
**Pull request by MikePresman**
If anyone has any ideas as to where else to add logs please let me know, this is temporary logging so i'd appreciate seeing more than less.

### CODE REVIEW

File: pkg/router/httpTriggers.go [hunk 1]
```diff
@@ -434,6 +435,7 @@ func (ts *HTTPTriggerSet) updateRouter(ctx context.Context) {
 			continue
 		}
 		ts.mutableRouter.updateRouter(router)
+		ts.logger.Info("updated router")
```
Comment (scott-rc) [inline pkg/router/httpTriggers.go]:
```suggestion
		ts.logger.Info("updated router", zap.Int("triggers", len(ts.triggers)), zap.Int("functions", len(ts.functions))
```

File: pkg/router/httpTriggers.go [hunk 2]
```diff
@@ -423,6 +423,7 @@ func (ts *HTTPTriggerSet) updateRouter(ctx context.Context) {
 				fn := *f.(*fv1.Function)
 				functionTimeout[fn.ObjectMeta.UID] = fn.Spec.FunctionTimeout
 				allfunctions = append(allfunctions, fn)
+				ts.logger.Info("allFunctions count %d", zap.Int("count", len(allfunctions)))
```
Comment (scott-rc) [inline pkg/router/httpTriggers.go]: All arguments after the message are key/value pairs, so this would end up logging:
```json
{ "message": "allFunctions count %d", "count": 123 }
```

I'd suggest dropping this log and just logging the trigger and function counts in the `updated router` log below.

File: pkg/router/functionServiceMap.go [hunk 3]
```diff
@@ -79,5 +83,7 @@ func (fmap *functionServiceMap) assign(f *metav1.ObjectMeta, serviceURL *url.URL
 
 func (fmap *functionServiceMap) remove(f *metav1.ObjectMeta) error {
 	mk := keyFromMetadata(f)
+
+	fmap.logger.Info("removing key and value from functionServiceMap")
```
Comment (scott-rc) [inline pkg/router/functionServiceMap.go]: Do you know how often these logs will be emitted?

I'm worried we're going to blow our Axiom ingest if it's very high 😅

Comment (MikePresman) [reply pkg/router/functionServiceMap.go]: No clue, I guess i'll remove the logs mutating the map

**Review by scott-rc — APPROVED** [review]

File: Makefile
Comment (scott-rc) [inline Makefile:1]: nit: I'd avoid changing this if we don't have to

**Comment by MikePresman** [comment]
Ran this locally and I see the logs, not much use locally as I want to use them to correlate to 404's !

File: pkg/router/functionServiceMap.go [hunk 4]
```diff
@@ -42,6 +42,8 @@ type (
 )
 
 func makeFunctionServiceMap(logger *zap.Logger, expiry time.Duration) *functionServiceMap {
+	logger.Info("creating functionServiceMap")
```
Comment (angelini) [inline pkg/router/functionServiceMap.go]: I don't understand the purpose of this log line?

What is it going to answer? You already log in `Start` and this function is only called during `Start`.

**Comment by angelini** [comment]
I think the most important log we're missing here is one when the router returns a 404, that's the bug we're chasing we should capture that data point.

File: pkg/router/functionHandler.go [hunk 5]
```diff
@@ -614,6 +614,7 @@ func (fh functionHandler) getServiceEntryFromCache() (serviceUrl *url.URL, err e
 			}
 			errMsg = fmt.Sprintf("Error getting function %v;s service entry from cache: %v", fh.function.ObjectMeta.Name, err)
 		}
+		fh.logger.Info("404 getting function", zap.String("error_message", errMsg))
```
Comment (angelini) [inline pkg/router/functionHandler.go]: This isn't a 404, right underneath it is returning `http.StatusInternalServerError` which is 500.

File: pkg/router/httpTriggers.go [hunk 6]
```diff
@@ -143,13 +144,21 @@ func versionHandler(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func NotFoundHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("404 Not Found: %s %s from %s at %s", r.Method, r.URL.Path, r.RemoteAddr, time.Now().Format(time.RFC3339))
+	w.WriteHeader(http.StatusNotFound)
+}
```
Comment (scott-rc) [inline pkg/router/httpTriggers.go]: This is going to be hard to parse in Axiom.

I suggest turning this into an anonymous function on line 160 and using the `ts.logger` so it's automatically serialized as json.

e.g.
```go
muxRouter.NotFoundHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
	ts.logger.Info("404 Not Found", zap.String("req_method", r.Method), /* etc... */)
	w.WriteHeader(http.StatusNotFound)
})
```

Try to re-use as many fields that as you can that we already index in Axiom.

e.g.
CleanShot 2024-05-23 at 14 01 51@2x

File: pkg/router/functionHandler.go [hunk 7]
```diff
@@ -660,7 +660,7 @@ func (fh functionHandler) getServiceEntryFromExecutor(ctx context.Context) (serv
 	// parse the address into url
 	svcURL, err := url.Parse(fmt.Sprintf("http://%v", service))
 	if err != nil {
-		logger.Error("error parsing service url",
+		logger.Info("error parsing service url",
```
Comment (scott-rc) [inline pkg/router/functionHandler.go]: I don't think these need to be changed, `logger.Error` logs should already show up in Axiom.

Only `Debug` and lower (e.g. `Trace`) are disabled atm. `Info` and above (e.g. `Warn`, `Error`, etc.) should already be enabled.

File: pkg/router/httpTriggers.go [hunk 8]
```diff
@@ -143,13 +143,23 @@ func versionHandler(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func NotFoundHandler(ts *HTTPTriggerSet) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ts.logger.Info("404 Not Found", zap.String("req_method", r.Method), zap.String("url_path", r.URL.Path), zap.String("req_address", r.RemoteAddr), zap.String("Time", time.Now().Format(time.RFC3339)), zap.Int("route_table_size", len(ts.functions)))
```
Comment (MikePresman) [inline pkg/router/httpTriggers.go]: CleanShot 2024-05-23 at 14 19 25@2x

Note the route_table_size matches the number of functions so I'm confident this will provide good insight. (I snake_cased it in this change)
